aws/resource_aws_connect_instance.go

```diff
@@ -0,0 +1,284 @@
+				Optional: true,
+				ForceNew: true,
+				Default:  connect.DirectoryTypeConnectManaged,
+				ValidateFunc: validation.StringInSlice([]string{
+					connect.DirectoryTypeSaml,
+					connect.DirectoryTypeConnectManaged,
+					connect.DirectoryTypeExistingDirectory,
+				}, false),
+			},
+			"directory_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(12, 12),
+			},
+			"instance_alias": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 64),
+			},
+			"inbound_calls_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"outbound_calls_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"early_media_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"contact_flow_logs_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"contact_lens_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"auto_resolve_best_voices": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"use_custom_tts_voices": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"created_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"service_role": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsConnectInstanceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).connectconn
+
+	input := &connect.CreateInstanceInput{
+		ClientToken:            aws.String(resource.UniqueId()),
+		IdentityManagementType: aws.String(d.Get("identity_management_type").(string)),
+		InstanceAlias:          aws.String(d.Get("instance_alias").(string)),
+		InboundCallsEnabled:    aws.Bool(d.Get("inbound_calls_enabled").(bool)),
+		OutboundCallsEnabled:   aws.Bool(d.Get("outbound_calls_enabled").(bool)),
+	}
+
+	if _, ok := d.GetOk("directory_id"); ok {
+		input.DirectoryId = aws.String(d.Get("directory_id").(string))
+	}
+
+	log.Printf("[DEBUG] Creating Connect Instance %s", input)
+
+	output, err := conn.CreateInstanceWithContext(ctx, input)
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading Connect Instance (%s): %s", d.Id(), err))
+	}
+
+	d.SetId(aws.StringValue(output.Id))
+
+	if _, err := waiter.InstanceCreated(ctx, conn, d.Id()); err != nil {
+		return diag.FromErr(fmt.Errorf("error waiting for Connect instance creation (%s): %s", d.Id(), err))
+	}
+
+	for att := range resourceConnectInstanceAttributesMapping {
```

Comment:
Im wondering if the attributes should not be handled in a separate resource, this is a separate api call either way(for read/write of attribute)

Reply:
Not sure I understood what you mean by that can you please explain?

Reply:
I'm not sure if the suggestion is correct, the attributes are dependent on the instance and breaking them out as separate resources would not be advised. The pattern with the connect API's is coarse-grained create with the fine-grained update.  I think we should stick to exposing the coarse-grained resource and hiding the chatter

Reply:
My thought is that another resource called aws_connect_instance_attribute as adding an attribute is a separate function call.
Similar example is splitting Efs file system and policy to individual resources https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/efs_file_system_policy.


Keeping these separate makes it easier for operators to define IAM policies and restrict permissions based on the actions the resource creation needs (in this case instance creation and attribute addition to the instance)

Reply:
I'm in agreement with @thornleyk - there isn't an API call to create/delete attributes but rather just a call to [UpdateInstanceAttribute](https://docs.aws.amazon.com/connect/latest/APIReference/API_UpdateInstanceAttribute.html) with the instance ID and the name of the attribute (which is from a pre-defined list). There's precedent for this in other AWS APIs - the EC2 API has a [ModifyInstanceAttribute](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyInstanceAttribute.html) API method for setting attributes of an EC2 instance for example - those attributes are mapped to appropriate Terraform attributes in the `aws_instance` schema